sklearn/pipeline.py

```diff
@@ -220,20 +235,44 @@ def _final_estimator(self):
 
     def _fit(self, X, y=None, **fit_params):
         self._validate_steps()
+        # Setup the memory
+        memory = self.memory
+        if memory is None:
+            memory = Memory(cachedir=None, verbose=0)
+        elif isinstance(memory, six.string_types):
+            memory = Memory(cachedir=memory, verbose=0)
+        elif not isinstance(memory, Memory):
+            raise ValueError('memory is either `str` or a `joblib.Memory`'
+                             ' instance, got: %s' % memory)
+        # Decorate _fit_transform_one to be able to use it
+        # with cache
+        fit_transform_one = memory.cache(_fit_transform_one)
+
         fit_params_steps = dict((name, {}) for name, step in self.steps
                                 if step is not None)
         for pname, pval in six.iteritems(fit_params):
             step, param = pname.split('__', 1)
             fit_params_steps[step][param] = pval
         Xt = X
-        for name, transform in self.steps[:-1]:
+        for step_idx, (name, transform) in enumerate(self.steps[:-1]):
             if transform is None:
                 pass
-            elif hasattr(transform, "fit_transform"):
-                Xt = transform.fit_transform(Xt, y, **fit_params_steps[name])
             else:
-                Xt = transform.fit(Xt, y, **fit_params_steps[name]) \
-                              .transform(Xt)
+                if memory.cachedir is None:
+                    # we do not clone when caching is disabled to preserve
+                    # backward compatibility and the possibility to use warm
+                    # started models in that case
+                    cloned_transformer = transform
+                else:
+                    cloned_transformer = clone(transform)
+                # Fit or load from cache the current transfomer
+                Xt, fitted_transformer = fit_transform_one(
+                    cloned_transformer, name, None, Xt, y,
+                    **fit_params_steps[name])
+                # Update the transformer if caching is enabled
+                if memory.cachedir is not None:
+                    transform.__dict__ = fitted_transformer.__dict__
```

Comment:
Could you please explain why we need to do this, and should not just set `self.steps[step_idx] = (name, fitted_transformer)`?

You should be aware that `__dict__` is not the only place that an estimator may store its attributes. I think this will work for all scikit-learn-internal estimators, but not necessarily for those implemented in other projects (e.g. written in Cython).

(Also, I note the inconsistency in naming between `transform` and `fitted_transformer`. Rename both to `transform` or `transformer`.)